### ISSUE TITLE
Propegating current span across thread boundaries in async application

### ISSUE THREAD

**Issue by marcoferrer** [DataDog/dd-trace-java#652]
We're using kotlin coroutines for implementing async services. One issue that we've run into is properly persisting the current span when the executed code resumes on a different thread.

There are execution hooks available that are meant for setup and tear down of thread locals which are meant for this usage.

Problem Im facing is not knowing the proper way to attach / detach a span to the current thread. The other issue that comes to mind is how to handle active span mutations that are not managed by the execution hook ie. some instrumented code is executed which creates a new child span.

**Comment by tylerbenson** [comment]
Hey @marcoferrer, welcome!

I'm a bit unclear... are you asking a question or reporting a problem?

The main instrumentation we have to handle async is the [java_concurrent](https://github.com/DataDog/dd-trace-java/tree/master/dd-java-agent/instrumentation/java-concurrent) instrumentation which instruments java `Executor` implementations.  You can look at that instrumentation for more details how it works.

Ideally we want it to work OOTB for kotlin coroutines as well, so feel free to suggest changes or submit a PR if you see a quick/easy solution.  I reviewed the [coroutine docs](https://github.com/Kotlin/kotlinx.coroutines/blob/master/docs/coroutine-context-and-dispatchers.md#thread-local-data) on the topic and it's not obvious to me what we would need to do.

If you see a bug, it might be helpful to provide a sample application with instructions on how to run and reproduce the issue.

**Comment by marcoferrer** [comment]
Sorry about the lack of clarity. I was more interested in finding out the proper way to manage spans across thread boundaries. I didnt realized the `Executors` were being instrumented as well and would explain a few things. 

We didnt start having issues with spans in coroutines until recently. Looking back now, I think it might have to do with the default executor implementation for coroutines switching from a `ForkJoinPool` to a custom solution. 

I did some digging into the executor instrumentation and added support for the new coroutine schedulers. The tests seem to work so I think this PR will take care of it. https://github.com/DataDog/dd-trace-java/pull/653 

Let me know your thoughts. I did want to increase the test coverage a bit if these changes are acceptable.
